Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,212 @@
++                    int vd = e.getByteValue();
++                    //Paper end
+                     int j = this.f.get(i);
+                     int k = this.c(i);
+ 
+@@ -430,7 +440,7 @@ public abstract class ChunkMapDistance {
+                             }
+ 
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
++                        this.a(i, k, this.c(j), k <= vd - 2, vd); //Paper ticket level pp-view-dist
+                     }
+                 }
+ 
+@@ -448,6 +458,7 @@ public abstract class ChunkMapDistance {
+ 
+         protected final Long2ByteMap a = new Long2ByteOpenHashMap();
+         protected final int b;
++        protected int pvd; //Paper ticket level pp-view-dist
+ 
+         protected b(int i) {
+             super(i + 2, 16, 256);
+@@ -455,6 +466,8 @@ public abstract class ChunkMapDistance {
+             this.a.defaultReturnValue((byte) (i + 2));
+         }
+ 
++        protected int getChunksMapping(long i) {return this.c(i);} //Paper OBFHELPER
++
+         @Override
+         protected int c(long i) {
+             return this.a.get(i);
+@@ -464,16 +477,34 @@ public abstract class ChunkMapDistance {
+         protected void a(long i, int j) {
+             byte b0;
+ 
++            //Paper start ticket level pp-view-dist
++            if(j == 0) {
++                ObjectSet<EntityPlayer> c = ChunkMapDistance.this.getChunkPlayers(i);
++                if (c != null && !c.isEmpty()) {
++                    ObjectIterator<EntityPlayer> pl = c.iterator();
++                    int[] vd = new int[c.size()];
++                    int x = 0;
++                    while (pl.hasNext()) {
++                        EntityPlayer p = pl.next();
++                        vd[x] = p.getViewDistance();
++                        java.util.Arrays.sort(vd);
++                        this.pvd = vd[c.size() - 1];
++                        x++;
++                    }
++                }
++            }
++            //Paper end
++
+             if (j > this.b) {
+                 b0 = this.a.remove(i);
+             } else {
+                 b0 = this.a.put(i, (byte) j);
+             }
+ 
+-            this.a(i, b0, j);
++            this.a(i, b0, j, (byte) this.pvd); //Paper ticket level pp-view-dist
+         }
+ 
+-        protected void a(long i, int j, int k) {}
++        protected void a(long i, int j, int k, byte d) {} //Paper ticket level pp-view-dist
+ 
+         @Override
+         protected int b(long i) {
+diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
+index cd73cde8d..0b38b2e67 100644
+--- a/src/main/java/net/minecraft/server/EntityHuman.java
++++ b/src/main/java/net/minecraft/server/EntityHuman.java
+@@ -80,6 +80,14 @@ public abstract class EntityHuman extends EntityLiving {
+     public String spawnWorld = "";
+     public int oldLevel = -1;
+ 
++    //Paper start ticket level pp-view-dist
```

Comment:
ideally we keep view distance for players who have not had it set by plugins to -1 given mojang's code supports changing view distance (globally) on the fly

Reply:
noted and changing

Reply:
actually the view distance is clamped and defaults to whatever value exists in playerChunkMap, preventing initial loads from failing

Reply:
rather not handicap us for implementing a potential World#setViewDistance given mojang's code clearly intends to support it (which this pr still does)